atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/ClientLogEvents.java

```diff
@@ -94,10 +101,54 @@ default CommitUpdate toCommitUpdate(
+
+        Set<LockDescriptor> locksFromLockWatchCreatedEvents = events().events().stream()
+                .flatMap(event -> event.accept(LockWatchCreatedEventVisitor.INSTANCE).stream())
+                .collect(Collectors.toUnmodifiableSet());
+        List<LockEvent> lockEvents = events().events().stream()
                 .flatMap(event -> event.accept(eventVisitor).stream())
-                .collect(ImmutableSet.toImmutableSet());
-        return CommitUpdate.invalidateSome(locksTakenOut);
+                .collect(Collectors.toUnmodifiableList());
+        Set<LockDescriptor> locksFromLockEvents = lockEvents.stream()
+                .map(LockEvent::lockDescriptors)
+                .flatMap(Set::stream)
+                .collect(Collectors.toUnmodifiableSet());
+
+        /*
+         * Metadata aggregation semantics for a lock descriptor L:
+         * - If there is a lock event that contains L, but L does not have metadata associated in this lock event, L is
+         *   excluded from the aggregation result.
+         * - Likewise, if L never shows up in a lock event (e.g., if it only appears in a LockWatchCreatedEvent), it is
+         *   also excluded from the aggregation result.
+         * - Otherwise, the metadata objects associated with L are listed in the order of the lock events they appear
+         *   in.
+         */
+        Map<LockDescriptor, ImmutableList.Builder<ChangeMetadata>> locksWithAggregatedMetadataBuilder = KeyedStream.of(
+                        locksFromLockEvents.stream())
+                .map(_unused -> new ImmutableList.Builder<ChangeMetadata>())
+                .collectToMap();
+        for (LockEvent lockEvent : lockEvents) {
+            for (LockDescriptor lock : lockEvent.lockDescriptors()) {
+                if (!locksWithAggregatedMetadataBuilder.containsKey(lock)) {
+                    continue;
+                }
+                Optional<ChangeMetadata> changeMetadata = lockEvent
+                        .metadata()
+                        .map(metadata ->
+                                metadata.lockDescriptorToChangeMetadata().get(lock));
+                if (lockEvent.metadata().isEmpty() || changeMetadata.isEmpty()) {
+                    locksWithAggregatedMetadataBuilder.remove(lock);
+                } else {
+                    locksWithAggregatedMetadataBuilder.get(lock).add(changeMetadata.get());
+                }
+            }
+        }
+
+        Set<LockDescriptor> locksTakenOut = Sets.union(locksFromLockEvents, locksFromLockWatchCreatedEvents);
+        Map<LockDescriptor, List<ChangeMetadata>> locksWithAggregatedMetadata = KeyedStream.ofEntries(
+                        locksWithAggregatedMetadataBuilder.entrySet().stream())
+                .<List<ChangeMetadata>>map(ImmutableList.Builder::build)
+                .collectToMap();
+        return CommitUpdate.invalidateSome(locksTakenOut, locksWithAggregatedMetadata);
```

Comment:
This is probably not the cleanest way for this computation, but it expresses the semantic quite clearly